File: pxr/usd/pcp/testenv/testPcpStreamingLayerReload.py
#!/pxrpythonsubst
#
# Copyright 2021 Pixar
#
# Licensed under the terms set forth in the LICENSE.txt file available at
# https://openusd.org/license.

from __future__ import print_function

from pxr import Sdf, Pcp, Plug, Vt
import os, unittest

class TestPcpStreamingLayerReload(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        testRoot = os.path.join(os.path.dirname(__file__), 'PcpPlugins')
        testPluginsDso = testRoot + '/lib'
        testPluginsDsoSearch = testPluginsDso + '/*/Resources/'

        # Register dso plugins.  Discard possible exception due to
        # TestPlugDsoEmpty.  The exception only shows up here if it happens in
        # the main thread so we can't rely on it.
        try:
            Plug.Registry().RegisterPlugins(testPluginsDsoSearch)
        except RuntimeError:
            pass

    def setUp(self):
        # We expect there to be no layers left loaded when we start each test
        # case so we can start fresh. By the tearDown completes this needs to 
        # be true.
        self.assertFalse(Sdf.Layer.GetLoadedLayers())

    def _CreatePcpCache(self, rootLayer):
        return Pcp.Cache(Pcp.LayerStackIdentifier(rootLayer))

    def test_StreamingLayerReload(self):

        # Open a layer with our streaming format.
        l = Sdf.Layer.FindOrOpen('root.testpcpstreaminglayerreload')
        self.assertTrue(l)

        # Build a cache.
        cache = self._CreatePcpCache(l)

        # Attempt to compute an index for /torus1/mesh_0 (should not exist).
        primIndex, errors = cache.ComputePrimIndex('/torus1/mesh_0')
        self.assertEqual(primIndex.primStack, [])

        # Load up asset.sdf, and replace l's content with it.  This only changes
        # the sublayer list, which pcp should recognize and blow layer stacks.
        # Since l's underlying data implementation returns true for
        # "StreamsData()" this exercises a different code-path in Pcp's change
        # processing.
        assetLayer = Sdf.Layer.FindOrOpen('asset.sdf')
        self.assertTrue(assetLayer)
        with Pcp._TestChangeProcessor(cache):
            l.TransferContent(assetLayer)

        # Now when we compute the index for the mesh, it should have a spec, due
        # to the added sublayer.
        primIndex, errors = cache.ComputePrimIndex('/torus1/mesh_0')
        self.assertTrue(len(primIndex.primStack) > 0)

if __name__ == "__main__":
    unittest.main()

File: pxr/usd/pcp/testenv/TestPcpStreamingLayerReload.cpp
//
// Copyright 2021 Pixar
//
// Licensed under the terms set forth in the LICENSE.txt file available at
// https://openusd.org/license.
//
#include "pxr/pxr.h"
#include "pxr/base/tf/staticTokens.h"
#include "pxr/usd/sdf/data.h"
#include "pxr/usd/sdf/textFileFormat.h"
#include "pxr/usd/sdf/layer.h"
#include "pxr/usd/sdf/textFileFormat.h"


PXR_NAMESPACE_OPEN_SCOPE

// Our SdfData subclass overrides StreamsData() to return true for testing.
class Test_PcpStreamingData : public SdfData
{
    virtual ~Test_PcpStreamingData();
    virtual bool StreamsData() const {
        return true;
    }
};

Test_PcpStreamingData::~Test_PcpStreamingData()
{
}

#define TEST_PCP_STREAMING_LAYER_RELOAD_TOKENS       \
    ((Id, "testpcpstreaminglayerreload"))            \
    ((Version, "1.0"))                               \
    ((Target, "usd"))

TF_DECLARE_PUBLIC_TOKENS(Test_PcpStreamingLayerReload_FileFormatTokens,
                         TEST_PCP_STREAMING_LAYER_RELOAD_TOKENS);

TF_DEFINE_PUBLIC_TOKENS(
    Test_PcpStreamingLayerReload_FileFormatTokens,
    TEST_PCP_STREAMING_LAYER_RELOAD_TOKENS);

/// \class Test_PcpStreamingLayerReload_FileFormat
///
class Test_PcpStreamingLayerReload_FileFormat : public SdfTextFileFormat
{
private:
    SDF_FILE_FORMAT_FACTORY_ACCESS;
    Test_PcpStreamingLayerReload_FileFormat();
    virtual ~Test_PcpStreamingLayerReload_FileFormat();

    virtual SdfAbstractDataRefPtr
    InitData(const FileFormatArguments &args) const;

};

TF_REGISTRY_FUNCTION(TfType)
{
    SDF_DEFINE_FILE_FORMAT(
        Test_PcpStreamingLayerReload_FileFormat, SdfTextFileFormat);
}

Test_PcpStreamingLayerReload_FileFormat
::Test_PcpStreamingLayerReload_FileFormat()
    : SdfTextFileFormat(
        Test_PcpStreamingLayerReload_FileFormatTokens->Id,
        Test_PcpStreamingLayerReload_FileFormatTokens->Version,
        Test_PcpStreamingLayerReload_FileFormatTokens->Target)
{
}

Test_PcpStreamingLayerReload_FileFormat
::~Test_PcpStreamingLayerReload_FileFormat()
{
}

SdfAbstractDataRefPtr
Test_PcpStreamingLayerReload_FileFormat
::InitData(const FileFormatArguments &args) const {
    SdfData *metadata = new Test_PcpStreamingData;
    // The pseudo-root spec must always exist in a layer's SdfData, so
    // add it here.
    metadata->CreateSpec(SdfPath::AbsoluteRootPath(), SdfSpecTypePseudoRoot);
    return TfCreateRefPtr(metadata);
}   

PXR_NAMESPACE_CLOSE_SCOPE
